### PR TITLE
Check extra package version

### LINKED ISSUE #1453: Add extra package version verification before running it
Check the version of packages that are not mandatory.

This causes issues that can have an infinity range of possible exceptions.

### PR DESCRIPTION
**Pull request by Kludex**
- Closes #1453

Changes:
- Vendorized `packaging.version` and `packaging._structures` modules.
- Use `packaging.version.parse` to get the version of the packages, and be able to compare them.
- Ignore vendorized package, and comparison conditionals on `coverage`.

Questions:
- Do we really want this?
- Isn't it cleaner to just include `setuptools` as dependency?
- Do we have alternatives here? Not checking the version is really an option?

### CODE REVIEW

File: uvicorn/_vendor/packaging/_structures.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        return True
+
+    def __ge__(self, other: object) -> bool:
+        return True
+
+    def __neg__(self: object) -> "NegativeInfinityType":
+        return NegativeInfinity
+
+
+Infinity = InfinityType()
```
Comment (ahopkins) [inline uvicorn/_vendor/packaging/_structures.py:56]: `float("inf")`?

**Comment by adriangb** [comment]
I think vendoring makes sense when it's unlikely that users would otherwise have the dependency installed. I don't think that's the case here, there's a good chance they already have packaging as a dependency. Is there another reason we'd want to vendor it?

If we did vendor, could we use the knowledge that we only need to support checking versions on certain packages to reduce the amount of code needed? Versioning is a tricky thing, and I think `packaging` tries to handle it in it's full complexity, but don't need that right?

**Comment by Kludex** [comment]
> I think vendoring makes sense when it's unlikely that users would otherwise have the dependency installed. I don't think that's the case here, there's a good chance they already have packaging as a dependency. Is there another reason we'd want to vendor it?

Well... https://github.com/encode/uvicorn/issues/1358

**Comment by adriangb** [comment]
I don't understand. Making a mistake and forgetting to include a dependency is a reason to vendor it? Sorry if I'm missing context

**Comment by Kludex** [comment]
This PR is vendorizing because of project history mainly. We tried to maintain the core `uvicorn` with minimal dependencies. I know it's a different case, but today, I think adding `asgiref` was a mistake, for example.

Should we add `packaging` as dependency instead?

Seems pretty uncool to add a mandatory dependency to check the version of the dependencies that are not even mandatory... Right?

**Comment by adriangb** [comment]
You have a good point. The situation is a bit different though, I think it's likely that `asgiref` would already exist in the dependency graph.

That said, can we solve the problem without vendoring of `packaging`? In other words, would a simple `__version__.split(".")[:3]` work for us? I realize this would fail for a _lot_ of version parsing, but if we're talking about 2 packages with known versioning schemes, maybe it works?

**Comment by Kludex** [comment]
Let's proceed with #1461

**Pull request closed without merging**